### PR TITLE
fix(tokens): Fixes the potential of deleting a working refresh_token when no new one is available

### LINKED ISSUE #404: save_refresh_token may delete valid refresh_token and break refresh functionality
**Describe the bug**
`save_refresh_token()` is called after the initial request. In our case of interest, the payload bring a `request_token` field (in the case of Google, this happens if `access_type=offline`.
Such `refresh_token` may not have an expiration time.

The problem is that when a refresh is requested, in `ensure_tokens_still_fresh()`, `save_refresh_token()` is called again with the new response. But the new response may **not** provide a new `refresh_token`.
Google [states](https://developers.google.com/identity/protocols/oauth2/web-server#obtainingaccesstokens)
> Your application should store both tokens in a secure, long-lived location that is accessible between different invocations of your application. 

`save_refresh_token()` disregard the existing and still valid `refresh_token` and replace it with false (since the response to a renewal does not contain it)

**To Reproduce**
1. Use Google OIDC with refresh token
```php
        add_filter( 'openid-connect-generic-auth-url', function( string $url ) {
            return $url . '&access_type=offline&prompt=consent';
        });
```
2. Connect
3. See from your logs that you're disconnect after 1h (the first call to`refresh_token`) 

**Expected behavior**
I think Google OIDC should work out of the box

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only this plugin is active.
- [x] This bug happens with a default WordPress theme active.
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
- PHP Version:  7.4
- WordPress Version: 3.9
- Plugin Version: 3.9.0
- Identity Provider: Google
- Relevant Plugin Settings:

### PR DESCRIPTION
**Pull request by drzraf**
### All Submissions:

* [x] Have you followed the [plugin Contributing guideline](https://github.com/oidc-wp/openid-connect-generi/blob/dev/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

### Changes proposed in this Pull Request:

`save_refresh_token()` disregard the existing and still valid `refresh_token` and replace it with false (since the response to a renewal does not contain it)

Closes #404

### How to test the changes in this Pull Request:

1. Use Google OIDC with refresh token
```php
        add_filter( 'openid-connect-generic-auth-url', function( string $url ) {
            return $url . '&access_type=offline&prompt=consent';
        });
```
2. Connect
3. See from your logs that, without this PR you're disconnect after 1h (the first call to`refresh_token`) 

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

Do not empty refresh_token if renewed access tokens do not come with new refresh_token.

### CODE REVIEW

**Comment by drzraf** [comment]
ping?

**Comment by drzraf** [comment]
We have been using this in -production to fix our GSuite 1h-disconnection problems and this has been confirmed to work.

**Comment by drzraf** [comment]
ping

**Comment by timnolte** [comment]
@drzraf so actually, in doing some digging on this subject to make sure that things are being handled properly, and to specs. I happened upon an actual reason for why you were having this issue. It has been on the table to quite some time to replace much of the core functionality with a standard Composer package. I happened upon this issue:

https://github.com/jumbojett/OpenID-Connect-PHP/issues/286

Which further took me here:

https://stackoverflow.com/a/65702100

Basically, what is happening is that once you authenticate the first time with Google you won't get a new refresh token un later requests without using `prompt=consent` or `approval_prompt=force` and `access_type=offline`. This is apparently not well documented with Google/GSuite. I'm not inclined to include this PR as I think there is more to the fix that is really required.

**Comment by timnolte** [comment]
An additional follow-up that I happened to come across this MU Plugin that is essentially an add-on for this plugin that addresses the Google refresh token issue.

https://gitlab.com/animalequality/wp-openid/-/blob/master/wp-openid.php

**Comment by drzraf** [comment]
> https://gitlab.com/animalequality/wp-openid/-/blob/master/wp-openid.php

I wrote this code, and having a filter to customize the request (adding `prompt=consent`) is nice **but** it doesn't free this plugin from the responsibility of behaving correctly by **not emptying** an existing/valid refresh token which is what this PR is about.

**Comment by timnolte** [comment]
@drzraf from what I see this isn't some clear problem specifically with this plugin as an issue with Google specifically. I didn't close the PR or issue I simply postponed it until I can find some very clear documentation on all of the handling and can test with other IDPs. The very fact the the code submitted has a comment regarding the scenario of what if the token was actually expired and the comment said "dunno" is not a very confident response to the proper way this should be handled.

All of the notes I made were in regards to doing actual research into this issue to ensure that the correct solution is implemented and not just something that appears to be a fox for 1 IDP or use case.

**Comment by czarny** [comment]
Hi it is not matter only of one IDP. AWS Cognito also does not have refresh tokens rotation and that is fine with spec. See https://datatracker.ietf.org/doc/html/rfc6749#section-6
> The authorization server **MAY** issue a new refresh token, in which case
   the client MUST discard the old refresh token and replace it with the
   new refresh token.
@drzraf I think you should update your pull request. Especially comments. If token is expired auth server will return error and plugin will end user session.

**Comment by drzraf** [comment]
Just to remind the actual use-case: My users connect to WP using GSuite. Without such a modification of the refresh-token handling, they get disconnected after one hour what is understandably problematic and frustrating. I didn't read the standards and whether IDP is in charge of disconnecting (IMHO identity providing <> session management/duration).

**Comment by timnolte** [comment]
> Just to remind the actual use-case: My users connect to WP using GSuite. Without such a modification of the refresh-token handling, they get disconnected after one hour what is understandably problematic and frustrating. I didn't read the standards and whether IDP is in charge of disconnecting (IMHO identity providing <> session management/duration).

@drzraf so it's very possible the refresh token issue you are seeing is due to a different bug that has already been fixed but it hasn't been released for this version of the plugin. If you install the version I linked to in the pinned issue about no more updates going to WP.org you might find that your 1 hour timeout issue is fixed.

https://github.com/forumone/openid-connect-wp-dist
